### PR TITLE
Migrating to ec2_instance

### PR DESCRIPTION
**Pull request by thiagoalmeidasa**
* Using ec2_instance module instead of ec2.
* Using cloud-init facts to check if boot process is complete
* Adding region setup for ec2 create and destroy

### CODE REVIEW

File: molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/converge.yml [hunk 1]
```diff
@@ -2,6 +2,14 @@
 - name: Converge
   hosts: all
   tasks:
+    - name: Wait for boot process to finish
```
Comment (mohit-kumar-sharma) [inline molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/converge.yml:5]: we already have a place for these kind of thing. which is prepare.yml

Comment (thiagoalmeidasa) [reply molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/converge.yml:5]: Thanks for the review. Should I move this block to `prepare.yml`?

Comment (mohit-kumar-sharma) [reply molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/converge.yml:5]: yes, i think so 
```
- name: Wait for cloud init to finish
      cloud_init_data_facts:
        filter: status
      register: res
      until: "res.cloud_init_data_facts.status.v1.stage is defined and not res.cloud_init_data_facts.status.v1.stage"
      retries: 100
      delay: 10
```

File: molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml [hunk 2]
```diff
@@ -73,53 +73,49 @@
         filters:
           name: "{{ item.image_name }}"
       loop: "{{ molecule_yml.platforms }}"
-      when: item.image is not defined
+      when: item.image_id is not defined
       register: ami_info
 
     - name: Create molecule instance(s)
-      ec2:
+      ec2_instance:
```
Comment (mohit-kumar-sharma) [inline molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:80]: is there any special reason we are replacing the ec2 module with ec2_instance?

Comment (thiagoalmeidasa) [reply molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:80]: `The ec2 module documentation should probably be updated to suggest using the ec2_instance module.`

From this issue: https://github.com/ansible/ansible/issues/64610#issuecomment-551924294

Comment (mohit-kumar-sharma) [reply molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:80]: makes sense, thanks for sharing

**Comment by thiagoalmeidasa** [comment]
Is something missing?

File: molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml [hunk 3]
```diff
@@ -73,53 +73,49 @@
         filters:
           name: "{{ item.image_name }}"
       loop: "{{ molecule_yml.platforms }}"
-      when: item.image is not defined
+      when: item.image_id is not defined
```
Comment (lod) [inline molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:76]: I'm not sure switching from `item.image` to `item.image_id` is worthwhile

If it is, it should probably be a separate pull request

It doesn't align with the documentation, it also hurts potential upgrading by requiring the molecule.yml file to be updated

File: molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml [hunk 4]
```diff
@@ -73,53 +73,49 @@
...
+        network:
+          assign_public_ip: true
       register: server
       loop: "{{ molecule_yml.platforms }}"
       loop_control:
         index_var: index
-      async: 7200
-      poll: 0
-
-    - name: Wait for instance(s) creation to complete
```
Comment (lod) [inline molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:103]: Removing the creation waits is a substantial change, this shifts the burden of waiting for the creation to the prepare file.

To me this is very non-obvious, at the end of the create step the instance should have been created.
At the very least, this change should have accompanying documentation updates.

I know the initial proposal used cloud_init_data_facts as a replacement wait process.
That suffers from being Linux specific, it doesn't support Windows or FreeBSD based systems.

File: molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml [hunk 5]
```diff
@@ -73,53 +73,49 @@
...
-        count_tag:
-          instance: "{{ item.name }}"
+        wait_timeout: 300
+        network:
+          assign_public_ip: true
       register: server
       loop: "{{ molecule_yml.platforms }}"
       loop_control:
         index_var: index
-      async: 7200
```
Comment (lod) [inline molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:100]: Shifting from async to wait will significantly slow down creation of many machines, anything above five.

I also feel that shifting from a timeout of 7200 to 300 should be called out separately. I have no idea why the original timeout was two hours, that seems obscene. But shifting from two hours to five minutes is a substantial change.

AWS documentation states store backed instances are provisioned in "usually less than 5 minutes".
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ComponentsAMIs.html#storage-for-the-root-device

People reported two hours required for Windows instances back in 2009
https://serverfault.com/questions/43884/speeding-up-launch-of-amazon-ec2-windows-instances
However my current job has required creating many Windows instances without anywhere near that level of delay

Terraform uses a timeout of ten minutes, this is a well tested time, I've only once had an issue with this timeout.
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance

File: molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml (hunk 3 above)
Comment (lod) [inline molecule_ec2/cookiecutter/{{cookiecutter.molecule_directory}}/{{cookiecutter.scenario_name}}/create.yml:76]: Shifting from `item.image` to `item.image_id` doesn't seem worthwhile.

It will require updates to the documentation (not included) and break compatibility with older molecule.yml files

I can see that it brings consistency with the AWS calls but that feels like an insufficient reason to break with the standard interface.

**Review by lod — CHANGES_REQUESTED** [review]
ec2_key also needs a region parameter, both to create and delete

This is non-trivial as the propose structure allows each instance to specify a region, potentially leading to multiple regions

It might be better to rebase this merge request so that it doesn't include the region elements, leave those to #19

**Comment by ssbarnea** [comment]
I am closing as it seems that original OP did not pursuit in fixing noted issues. Please feel free to reopen.
